Apache httpasyncclient 5.x

anuragagarwal561994 opened the pull request:
Closes: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/5631

anuragagarwal561994 commented:
@ok2c @trask this is the first draft, let us just continue our discussion if any on this MR.

I am open to any suggestion or changes that are required, all the test cases are passing currently and I have also incorporated the test case changes that were suggested to me.

mateuszrzeszutek commented on docs/supported-libraries.md:
```diff
@@ -29,6 +29,7 @@ These are the supported libraries and frameworks:
 | [Apache CXF JAX-WS](https://cxf.apache.org/)                                                                                      | 3.0+                           |
 | [Apache Dubbo](https://github.com/apache/dubbo/)                                                                                  | 2.7+ (not including 3.x yet)   |
 | [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                        | 4.1+                           |
+| [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                        | 5.0+                           |
```
Can we merge these two entries?
```suggestion
| [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                        | 4.1+ and 5.0+                  |
```

mateuszrzeszutek commented on instrumentation/apache-httpasyncclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpAsyncClientHttpAttributesGetter.java:
```diff
@@ -0,0 +1,76 @@
+  @Override
+  @Nullable
+  public Integer statusCode(ApacheHttpClientRequest request, HttpResponse response) {
+    return response != null ? response.getCode() : null;
```
`response` will always be non-null here

mateuszrzeszutek commented on instrumentation/apache-httpasyncclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpClientRequest.java:
```diff
@@ -0,0 +1,127 @@
+    return uri != null ? uri.toString() : null;
+  }
+
+  public String getFlavor(HttpResponse response) {
```
Nit
```suggestion
  public String getFlavor(@Nullable HttpResponse response) {
```

anuragagarwal561994 replied:
This was actually done for interim purpose, actually if we see this is kind of a static method or more related to response and not the request, what do you suggest I should do here. This should ideally not be part of the AsyncHttpClientRequest now as it is now not a request property.

anuragagarwal561994 replied:
I will need to add a logger separately in the AttributesGetter as well that should be fine right?

mateuszrzeszutek commented on instrumentation/apache-httpasyncclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpClientRequest.java:
```diff
@@ -0,0 +1,127 @@
+
+public final class ApacheHttpClientRequest {
+
+  private static final Logger logger = LoggerFactory.getLogger(ApacheHttpClientRequest.class);
```
Recently we've removed all slf4j usage from most of the agent code and switched to JUL -- can you rebase against a fresh `main` and replace slf4j with `java.util.logging` logger?

anuragagarwal561994 replied:
done

mateuszrzeszutek commented on instrumentation/apache-httpasyncclient-5.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpAsyncClientTest.java:
```diff
@@ -0,0 +1,240 @@
+          attributes.add(SemanticAttributes.HTTP_URL);
+          attributes.add(SemanticAttributes.HTTP_METHOD);
+          attributes.add(SemanticAttributes.HTTP_USER_AGENT);
+          attributes.add(SemanticAttributes.HTTP_SCHEME);
+          attributes.add(SemanticAttributes.HTTP_TARGET);
```
`http.scheme` and `http.target` are not captured by the `HttpClientAttributesExtractor`, these are server-only
```suggestion
```

mateuszrzeszutek reviewed (commented):
Left a couple of comments, but overall it looks really good 👍 
Thanks!

anuragagarwal561994 commented:
Thanks @mateuszrzeszutek there are couple of things I would like to highlight with respect to the MR.
In HttpClient 4, we know while the instrumentation is initiated the flavor of the request.
In HttpClient 5, we know this after we have made the connection, which is after the instrumentation has started.

Thus in test cases where we are having issues connecting with the server, my tests were failing and thus for the test cases I have removed the check for flavor, although it still comes in for the successful request cases. Suggest me if I can do something here, as far as I saw the library I couldn't think of anything else.

anuragagarwal561994 commented:
Support for method

```
public final <T> Future<T> execute(
            final HttpHost target,
            final AsyncRequestProducer requestProducer,
            final AsyncResponseConsumer<T> responseConsumer,
            final HandlerFactory<AsyncPushConsumer> pushHandlerFactory,
            final HttpContext context,
            final FutureCallback<T> callback) {
    }
```

in `CloseableHttpAsyncClient` is not added and the test cases were not written for the same although a user might be using the same, because I thought instrumenting an interface would be right than instrumenting an abstract class implementing that interface.

So the instrumentation kind of doesn't cover all the cases but most of the cases for now.

anuragagarwal561994 commented:
Third thing that I would want to highlight here is that HTTP client 5 by default in Negotiate mode which means if http endpoint is given it will do http 1.1 but if https is given then it will prefer http 2.

Most of the tests were passed but `io.opentelemetry.instrumentation.testing.junit.http.AbstractHttpClientTest#httpsRequest` was failing as I think the server was returning http 1.1 response while the client was trying to parse it as http 2 or something happened here. If you help me with that case that would be great.

For now I enforced the http client 5 to use `HttpVersionPolicy.FORCE_HTTP_1` in tests to overcome this scenario, but if there is something we can do here, do let me know.

mateuszrzeszutek commented:
> Thus in test cases where we are having issues connecting with the server, my tests were failing and thus for the test cases I have removed the check for flavor, although it still comes in for the successful request cases. Suggest me if I can do something here, as far as I saw the library I couldn't think of anything else.

The `HttpClientTestOptions#setHttpAttributes()` method accepts a lambda that gets the URI called as a parameter -- depending on which endpoint was called (e.g. containing `/success`) you can decide whether to expect `http.flavor` or not.

> So the instrumentation kind of doesn't cover all the cases but most of the cases for now.

It can always be added in a separate PR, I believe it is fine for now (as long as we don't lose that information).

> For now I enforced the http client 5 to use `HttpVersionPolicy.FORCE_HTTP_1` in tests to overcome this scenario, but if there is something we can do here, do let me know.

We don't have any HTTP 2 test coverage yet (see #3227) -- for now you don't have to worry about, forcing HTTP/1.1 if perfectly fine.

anuragagarwal561994 commented:
I have addressed all the comments, @mateuszrzeszutek you can check the updated code

anuragagarwal561994 commented:
>The HttpClientTestOptions#setHttpAttributes() method accepts a lambda that gets the URI called as a parameter -- depending on which endpoint was called (e.g. containing /success) you can decide whether to expect http.flavor or not.

This is done as well, any other changes required?

trask commented on instrumentation/apache-httpasyncclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpAsyncClientInstrumentationModule.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.List;
+
+@AutoService(InstrumentationModule.class)
+public class ApacheHttpAsyncClientInstrumentationModule extends InstrumentationModule {
+  public ApacheHttpAsyncClientInstrumentationModule() {
+    super("apache-httpasyncclient", "apache-httpasyncclient-5.0");
+  }
```
this class would go away when merging into apache-httpclient-5.0

trask reviewed (commented):
thx @anuragagarwal561994!

anuragagarwal561994 commented:
Sure @trask let me try doing the same and check if some other issues come up. How about I keep the package name separate and merge the modules?

anuragagarwal561994 commented:
@trask I tried merging the modules, there is just one issue httpclient-5.0 is having groovy test cases as of now and I am not able to get it running on my local, may be things are done correctly just that I was not able to test in my IDE.

If you can suggest something it would be of help, it just either reports some imports missing or so.

On running gradle build in terminal I get
```
Execution failed for task ':instrumentation:grpc-1.6:testing:generateProto'.
> Could not resolve all files for configuration ':instrumentation:grpc-1.6:testing:protobufToolsLocator_protoc'.
   > Could not find protoc-3.3.0-osx-aarch_64.exe (com.google.protobuf:protoc:3.3.0).
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/3.3.0/protoc-3.3.0-osx-aarch_64.exe
```

this is irrespective of this change, just that I couldn't test it on my local.

In IDE while building I keep getting javax.annotation.Nullable not found

mateuszrzeszutek commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpAsyncClientSingletons.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.hc.core5.http.HttpResponse;
+
+public final class ApacheHttpAsyncClientSingletons {
+  private static final String INSTRUMENTATION_NAME = "io.opentelemetry.apache-httpasyncclient-5.0";
```
The gradle module has changed, so let's use the same instrumentation name as the sync client instrumentation:
```suggestion
  private static final String INSTRUMENTATION_NAME = "io.opentelemetry.apache-httpclient-5.0";
```

mateuszrzeszutek commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentationModule.java:
```diff
@@ -21,6 +20,7 @@ public ApacheHttpClientInstrumentationModule() {
   @Override
   public List<TypeInstrumentation> typeInstrumentations() {
-    return singletonList(new ApacheHttpClientInstrumentation());
+    return List.of(
+        new ApacheHttpClientInstrumentation(), new ApacheHttpAsyncClientInstrumentation());
```
The javaagent is compiled against Java 8, unfortunately you can't use Java 9+ methods/features:
```suggestion
    return Arrays.asList(
        new ApacheHttpClientInstrumentation(), new ApacheHttpAsyncClientInstrumentation());
```

trask commented:
> @trask I tried merging the modules, there is just one issue httpclient-5.0 is having groovy test cases as of now and I am not able to get it running on my local, may be things are done correctly just that I was not able to test in my IDE.

checked out your PR locally, pushed one small change back to your PR, and tests are passing locally for me

trask commented on docs/supported-libraries.md:
```diff
@@ -28,7 +28,7 @@ These are the supported libraries and frameworks:
 | [Apache CXF JAX-WS](https://cxf.apache.org/)                                                                                      | 3.0+                           |
 | [Apache Dubbo](https://github.com/apache/dubbo/)                                                                                  | 2.7+ (not including 3.x yet)   |
-| [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                        | 4.1+                           |
+| [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                        | 4.1+ and 5.0+                  |
```
I think remove this change (5.0+ is covered by 4.1+ anyways)

anuragagarwal561994 replied:
done

trask commented on instrumentation/apache-httpasyncclient-5.0/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,15 @@
+plugins {
```
can delete this file now

anuragagarwal561994 replied:
done

trask commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/async/ApacheHttpAsyncClientHttpAttributesGetter.java:
```diff
@@ -0,0 +1,101 @@
+
+  @Override
+  @Nullable
+  public String flavor(ApacheHttpClientRequest request, @Nullable HttpResponse response) {
```
if this is exactly the same as method in classic instrumentation, how about extracting common helper somewhere?

anuragagarwal561994 replied:
More or less it is same but the request type is different, I guess we will have to keep 2 classes.

trask commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/async/ApacheHttpClientRequest.java:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  // minimize memory overhead by not using streams
+  static List<String> headersToList(Header[] headers) {
```
same with this method, if exactly the same as method in classic

anuragagarwal561994 replied:
The suggestions are good, but I guess it will involve more refactoring in both parts of the client support.
Like when I started this MR, I thought of just using HttpRequest only which is http client specific, but then I got a use-case of keeping the extra class AsyncHttpClientRequest which was to store the entity.

Both the async and the http client are kind of using different strategies. In async we are decorating, in classic we are replacing the request instance. So better way would be to have something in common and another way could be to have something like Request / Response adapters which operates on HttpRequest / HttpResponse. I see other methods which repeat like getFlavor

anuragagarwal561994 commented:
@trask @mateuszrzeszutek I have done some rework on my pull request and now I guess it should look awesome 😄 

Let me summarise the ChangeLog:
* Removed the `ApacheHttpAsyncClientRequest` - only reason left to keep it was to get the request content length, for starters we can do away with it.
* Made everything generic to `HttpRequest` - this allowed reusing the same logic for async client and classic client
* Moved all the logic from respective attribute getters to `ApacheHttpClientUtils`, generally by design I don't prefer utility classes (because this is Java not a scripting language) except in cases where we are dealing with third party entities
* After 2nd and 3rd refactoring, only thing left in the async client package was just the instrumentation class, so merged both the packages, all of the code became reusable by this point
* For the test cases I moved the current http client test cases to java (another pain point resolved 😄)
* Moved the boilerplate code into `AbstractApacheHttpClientTest` class thus making both of them look similar.
* Used the `connectionTimeout` and `readTimeout` method where possible
* Where I couldn't use the timeout method, I used them as static constants.

anuragagarwal561994 commented:
@mateuszrzeszutek @trask did you get time to review this MR?

trask commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/ApacheHttpClientHelper.java:
```diff
@@ -3,9 +3,9 @@
  */
 
-package io.opentelemetry.javaagent.instrumentation.apachehttpclient.v5_0;
+package io.opentelemetry.javaagent.instrumentation.apachehttpclient;
```
let's keep `.v5_0` package, since it differentiates from `.v4_0` and `.v2_0`

anuragagarwal561994 replied:
done

trask reviewed (commented):
thx this is looking great! will do a more thorough review soon

trask commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v5_0/ApacheHttpClientHelper.java:
```diff
@@ -3,9 +3,9 @@
  */
 
-package io.opentelemetry.javaagent.instrumentation.apachehttpclient.v5_0;
+package io.opentelemetry.javaagent.instrumentation.v5_0;
```
i've had this Intellij refactoring problem before too 😄
```suggestion
package io.opentelemetry.javaagent.instrumentation.apachehttpclient.v5_0;
```

anuragagarwal561994 replied:
done

mateuszrzeszutek reviewed (approved):
Left a couple of minor comments, but overall LGTM 👍

anuragagarwal561994 commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java:
```diff
@@ -0,0 +1,267 @@
+        return;
+      }
+
+      ApacheHttpClientSingletons.instrumenter()
```
done

anuragagarwal561994 commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java:
```diff
@@ -130,11 +129,11 @@ public static void methodEnter(
         @Advice.Local("otelScope") Scope scope) {
       Context parentContext = currentContext();
-      if (!instrumenter().shouldStart(parentContext, request)) {
+      if (!ApacheHttpClientSingletons.instrumenter().shouldStart(parentContext, request)) {
```
done

trask commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientUtils.java:
```diff
@@ -0,0 +1,146 @@
+import org.apache.hc.core5.http.ProtocolVersion;
+import org.apache.hc.core5.net.URIAuthority;
+
+class ApacheHttpClientUtils {
```
the methods in this class look like they are not really shared (each only called from one place), maybe inline them into the callers (or move the methods as private methods in the classes where they are used)?

anuragagarwal561994 replied:
done

anuragagarwal561994 commented on instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/WrappingStatusSettingResponseHandler.java:
```diff
@@ -34,7 +32,7 @@ public WrappingStatusSettingResponseHandler(
   @Override
   public T handleResponse(ClassicHttpResponse response) throws IOException, HttpException {
-    instrumenter().end(context, request, response, null);
+    ApacheHttpClientSingletons.instrumenter().end(context, request, response, null);
```
done

trask reviewed (approved):
thx @anuragagarwal561994 💯
